### ISSUE TITLE
Text Generation Stream not cleaning up properly on AbortSignal

### ISSUE THREAD

**Issue by swarupsrini** [huggingface/text-generation-inference#1199]
### System Info

Running the publicly available image on GHCR on an ubuntu host:
```
sudo docker run --name llmserver --gpus all --shm-size 1g -p 8080:80 -e CARGO_LOG="debug" -v /home/user/data:/data ghcr.io/huggingface/text-generation-inference:1.3 --model-id <model_id_here>  --quantize awq --max-batch-prefill-tokens 1500 --max-total-tokens 1300 --max-input-length 1200 --max-batch-total-tokens 1500
```

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

Follow up to https://github.com/huggingface/text-generation-inference/issues/1130. We use streaming for text generation and I implemented an AbortSignal to drop the request under custom stopping criteria. 

However, I noticed a few things:
1. The `INFO` log doesn't show up for the aborted stream.
2. I don't think Prometheus metrics (e.g. `tgi_request_count`) are updated either.
3. Most importantly, after a couple of aborted requests, we see these two errors consistently for the following stream requests:
    - Out of available cache blocks error: https://gist.github.com/swarupsrini/325355eb512a83b848122a7ade09e15b
    - CUDA out of memory error: https://gist.github.com/swarupsrini/08e5e818274d5cbc7a3fbd922503fc5b

### Expected behavior

Semantically, I understand that an aborted connection seems like an incomplete request, but for our use case it is complete. I think there's no alternative for us right now if we want custom stopping logic (even regex would suffice but the server doesn't seem to support that). 

I think most important are the errors we see consistently after a couple of aborted requests. My suspicion is that perhaps the proper clean-up process after each completed request was not done for any stream requests that were aborted, so I'm filing this issue in that case. 

Happy to hear any thoughts/suggestions!

**Comment by swarupsrini** [comment]
Bumping this since I think it's still a bug in the latest version. I think the issue could be in the `generate_stream` method - I'm suspect that parts of the clean up code aren't triggered during an abort signal.

https://github.com/huggingface/text-generation-inference/blob/ec6d4592d59c136ea1d2cac3b4e2a7ea004a0ce3/router/src/server.rs#L333

**Comment by swarupsrini** [comment]
We ran a script that submitted 50 requests of 1600 tokens each and aborts mid-way, but this didn't cause any errors. However, when we deploy to production, we see these errors for most incoming requests after an unpredictable amount of time (~5 min - 4 hrs).

**Comment by swarupsrini** [comment]
@OlivierDehaene tagging since you worked on this portion the most. Sorry for the ping, but I did do a lot of testing and think this is a bug in the code so would appreciate your thoughts here!

**Comment by OlivierDehaene** [comment]
>  Just ran some more tests, and got this error after a couple of streams with abort signal sent.

Do you get the same error when not aborting?

I don't think aborting is the issue as the cleanup code is exactly the same when aborting or not. It's possible that the issue comes from a miscalculation of the total free memory at warmup. 

Can you try with `MEMORY_FRACTION=0.8` for example? 
What model is it? Can you share the output of `--env`?

**Comment by swarupsrini** [comment]
> Do you get the same error when not aborting?

Not at all. We had our model deployed for ~2 months now without any errors and day I tried to use abort signals again, I saw this error after ~10 minutes. So I'd be quite surprised if this is not related to the aborting. 

I did notice that we don't see the SUCCESS log for any of the aborted requests, and it shows up as a failed 'dropped' request in the metrics, so I'm wondering if a lot of these aborted requests is causing any memory leaks.

> Can you try with `MEMORY_FRACTION=0.8` for example? 

I set `MEMORY_FRACTION=0.9` this morning and turned aborts back on, and we haven't seen any issues _yet_. Will report back if any issues! So your miscalculation theory makes sense, but it is strange since the warmup calculation should be independent of aborted requests.

> What model is it? Can you share the output of `--env?`

https://gist.github.com/swarupsrini/10aec0163ed8e44359c254f691c87e32
